CNV-49608:  Default to raw VM in VirtualMachineNavPageTitle

rszwajko opened the pull request:
## 📝 Description

Before the fix, VirtualMachineNavPageTitle used the result of
expand-spec instead of a raw VM object. The approach failed
in scenarios where:
1. isInstanceTypeVM() was used as this information is by design
   erased by the backend.
2. the VM was sent to the backend - as expanded VM is considered a new
   VM (with references to old VM removed)
The main section distinguished both objects and passed them separately
as 'instanceTypeExpandedSpec' and 'vm' (both of type V1VirtualMachine).

After the fix the raw VM object will be used by default in
VirtualMachineNavPageTitle and the expanded VM will be passed as
'instanceTypeExpandedSpec'.

Key points:
1. hasOwnProperty() checks used for 'autoattachGraphicsDevice' were
   refactored to use isHeadlessMode() helper
3. created an independent version of InstanceTypeModal that retrieves
   all available instance types - the extra query will be fired only
   when the modal is opened
4. retrieve the instance type name used by VirtualMachineInstance
   from annotations

Reference-Url: https://github.com/kubevirt-ui/kubevirt-plugin/pull/2414
Reference-Url: https://github.com/kubevirt/kubevirt/blob/47d69dd7dbf57932cbc70c59bab6e930d26cda3d/staging/src/kubevirt.io/api/core/v1/types.go#L1060
Reference-Url: https://github.com/kubevirt/kubevirt/blob/4c2642b9101d9b734a906ccc2dd157ca6276cf69/pkg/instancetype/expand/expand.go#L114

## 🎥 Demo

### Before
![Screenshot From 2025-03-06 21-04-38](https://github.com/user-attachments/assets/a3bb6bd8-cd27-4d67-b45c-c4d764ca67d4)
### After
![Screenshot From 2025-03-06 21-35-13](https://github.com/user-attachments/assets/aee46508-dd8b-401b-a75c-c47bd7f5ea7c)
![Screenshot From 2025-03-17 19-37-54](https://github.com/user-attachments/assets/d41d2a0f-c47e-40ae-89fe-0f4c01431bbc)

avivtur commented on src/utils/components/HardwareDevices/HardwareDevicesHeadlessMode.tsx:
```diff
@@ -16,9 +16,7 @@ const HardwareDevicesHeadlessMode: FC<HardwareDevicesHeadlessModeProps> = ({ onS
-  const [isChecked, setIsChecked] = useState<boolean>(
-    devices?.hasOwnProperty('autoattachGraphicsDevice') && !devices?.autoattachGraphicsDevice,
-  );
+  const [isChecked, setIsChecked] = useState<boolean>(devices?.autoattachGraphicsDevice === false);
```
I think isHeadlessModeVMI(vm?.spec?.template) should work the same, can you check?

metalice replied:
u can use isHeadlessModeVMI ?

rszwajko replied:
The helper has been made more generic and moved to `selectors.ts`

avivtur commented on src/utils/components/HeadlessMode/HeadlessMode.tsx:
```diff
@@ -10,10 +10,7 @@ type HeadlessModeProps = {
-    devices?.hasOwnProperty('autoattachGraphicsDevice') &&
-      devices?.autoattachGraphicsDevice === false,
-  );
+  const [isChecked, setIsChecked] = useState<boolean>(devices?.autoattachGraphicsDevice === false);
```
same here if isHeadlessModeVMI(vm?.spec?.template) works the same please change to utilize the util function :)

avivtur commented on src/utils/components/InstanceTypeModal/InstanceTypeModal.tsx:
```diff
@@ -119,4 +122,26 @@ const InstanceTypeModal: FC<InstanceTypeModalProps> = ({
   );
 };
 
+export const ConnectedInstanceTypeModal: FC<
```
we don't put 2 components into one file, please move this to a different file

avivtur commented on src/utils/components/InstanceTypeModal/InstanceTypeModal.tsx:
```diff
@@ -119,4 +122,26 @@ const InstanceTypeModal: FC<InstanceTypeModalProps> = ({
 
+export const ConnectedInstanceTypeModal: FC<
+  Pick<InstanceTypeModalProps, 'instanceTypeVM' | 'isOpen' | 'onClose' | 'onSubmit'>
+> = (props) => {
```
if props are used in the component and not just passed on, we should destruct if for readability

avivtur commented on src/utils/resources/vm/utils/instanceTypes.ts:
```diff
@@ -1,5 +1,19 @@
 export const isInstanceTypeVM = (vm: V1VirtualMachine): boolean =>
   !isEmpty(vm?.spec?.instancetype) || !isEmpty(vm?.spec?.preference);
+
+const NAMESPACED_INSTANCE_TYPE_NAME_ANNOTATION = 'kubevirt.io/instancetype-name';
+const CLUSTER_INSTANCE_TYPE_NAME_ANNOTATION = 'kubevirt.io/cluster-instancetype-name';
+const NAMESPACED_PREFERENCE_NAME_ANNOTATION = 'kubevirt.io/preference-name';
+const CLUSTER_PREFERENCE_NAME_ANNOTATION = 'kubevirt.io/cluster-preference-name';
```
Please use separated files for constants and utils

rszwajko replied:
Used existing (empty) `src/utils/resources/instancetype/constants.ts` file.

metalice commented on src/utils/components/HardwareDevices/modal/HardwareDevicesHeadlessModeModal.tsx:
```diff
@@ -26,9 +26,7 @@ const HardwareDevicesHeadlessModeModal: FC<HardwareDevicesHeadlessModeModalProps
-  const [checked, setChecked] = useState<boolean>(
-    devices?.hasOwnProperty('autoattachGraphicsDevice') && !devices?.autoattachGraphicsDevice,
-  );
+  const [checked, setChecked] = useState<boolean>(devices?.autoattachGraphicsDevice === false);
```
u can use isHeadlessModeVMI ? expand it to also include vm

metalice commented on src/views/virtualmachines/details/tabs/configuration/details/components/DetailsSectionBoot.tsx:
```diff
@@ -49,7 +47,7 @@ const DetailsSectionBoot: FC<DetailsSectionBootProps> = ({
   useEffect(() => {
     expandURLHash(getSearchItemsIds(getDetailsTabBootIds(vm)), location?.hash, setIsExpanded);
-  }, [vm, location?.hash]);
+  }, [vm, location?.hash, setIsExpanded]);
```
no need to include setState into deps - do u have extensions that announce if some deps are missing?

rszwajko replied:
The method comes from our custom hook `useToggle()`  - if it's not added to the list I get warnings in IDE

metalice commented on src/utils/components/PendingChanges/utils/helpers.ts:
```diff
@@ -444,17 +455,12 @@ export const getChangedHeadlessMode = (
-  const vmHeadless = !!vmDevices?.autoattachGraphicsDevice;
-  const vmiHeadless = !!vmiDevices?.autoattachGraphicsDevice;
+  // assume headless only if the property is explicitly false
+  const vmHasGraphic = vm?.spec?.template?.spec?.domain?.devices?.autoattachGraphicsDevice ?? true;
```
isHeadlessModeVMI can be used here?

rszwajko commented on src/utils/resources/vm/hooks/useInstanceTypeExpandSpec.ts:
```diff
@@ -35,7 +35,16 @@ const useInstanceTypeExpandSpec: UseInstanceTypeExpandSpec = (vm) => {
+          // re-establish the reference to the instance type
+          // expand-spec removes it as the output should allow creating an independent VM
+          instancetype: innerVM.spec.instancetype,
+          preference: innerVM.spec.preference,
```
Switched to annotations.

rszwajko commented:
@metalice @avivtur 
Reverted back to passing expanded VM and raw VM separately. 
Note that I've also dropped  the usage of ` V1VirtualMachineFromJSON(json)`.
Unfortunately the YAML editor is not supporting properties set to `undefined`.
![image](https://github.com/user-attachments/assets/00a13943-ad9a-4661-a3cc-87bbd639c42e)

metalice commented on src/utils/components/InstanceTypeModal/ConnectedInstanceTypeModal.tsx:
```diff
@@ -0,0 +1,33 @@
+      {loaded && loadError && <ErrorAlert error={loadError} />}
+      {loaded && !loadError && (
+        <InstanceTypeModal
+          {...otherProps}
```
as a rule of thumb , using {...otherProps} , should be avoided.
For example, lets say other props contiant 3 props.
if InstanceTypeModal , has useMemo, useCallback, that listen to only 1 prop from otherProps, it will recalc even if the 2 other props are getting new values, as {...otherProps} will always create a new object, new reference so useMemo and useCallback will fail compare and recalc. better to explicit pass props.

rszwajko replied:
I will check it but we should be OK here. The component receives individual props which have the same references.

rszwajko replied:
Works for me.

metalice commented on src/utils/components/InstanceTypeModal/ConnectedInstanceTypeModal.tsx:
```diff
@@ -0,0 +1,33 @@
+
+import InstanceTypeModal, { InstanceTypeModalProps } from './InstanceTypeModal';
+
+const ConnectedInstanceTypeModal: FC<
```
The name of this compent isnt clear, what does it mean connected? Please try rename to something clearer

rszwajko replied:
The naming comes from  React Redux - in this context it's just that (some) data is connected to the component.
We can change it - what do you propose?

metalice replied:
I will call this InstanceTypeModal and the inner component InstanceTypeModalContent or some variation

rszwajko replied:
In the end I've used `StandaloneInstanceTypeModal`. The `InstanceTypeModalContent` was not accurate because both components are used directly.

metalice commented on src/utils/components/InstanceTypeModal/ConnectedInstanceTypeModal.tsx:
```diff
@@ -0,0 +1,33 @@
+  );
+  return (
+    <>
+      {!loaded && <Loading />}
+      {loaded && loadError && <ErrorAlert error={loadError} />}
+      {loaded && !loadError && (
```
i think it more clear to work with returns

if (!loaded) return Loaidng
...

metalice commented on src/utils/components/PendingChanges/hooks/usePendingChanges.tsx:
```diff
@@ -80,6 +84,7 @@ export const usePendingChanges = (
     isList: true,
   });
 
+  const instanceTypeChanged = checkInstanceTypeChanged(vm, vmi);
```
You introduce new IT change, what happen if u change IT? We are getting the message or IT change and memory and CPU? both of them?

rszwajko replied:
> what happen if u change IT?
>
notification is displayed - only for the instance type (as visible on the screenshot)

metalice commented on src/utils/components/InstanceTypeModal/InstanceTypeModal.tsx:
```diff
@@ -17,7 +17,7 @@ import {
 } from './utils/util';
 
-type InstanceTypeModalProps = {
+export type InstanceTypeModalProps = {
```
if u are exporting type, please extract it to a util/type file

metalice commented on src/utils/components/InstanceTypeModal/StandaloneInstanceTypeModal.tsx:
```diff
@@ -0,0 +1,36 @@
+  Pick<InstanceTypeModalProps, 'isOpen' | 'onClose' | 'onSubmit' | 'vm'>
+> = ({ vm, ...otherProps }) => {
+  const { allInstanceTypes, loaded, loadError } = useInstanceTypesAndPreferences();
+  const instanceType = allInstanceTypes.find(
```
shouldnt this be on useMemo? as u are passing all items each rendering

rszwajko replied:
IMHO it makes no difference. It could be even slower  - `allInstanceTypes` is less then 20 items. We can add it if you prefer.

metalice replied:
They are going to introduce new families of IT - so better include it as potential this number can be much larger

metalice commented on src/utils/components/InstanceTypeModal/utils/util.ts:
```diff
@@ -80,34 +83,42 @@ const sortInstanceTypeSizes = (a: InstanceTypeRecord, b: InstanceTypeRecord) =>
-    (size) => size.prettyDisplaySize === instanceTypeSize,
+
+  if (!matchedSeries) {
+    return undefined;
```
undefined isn't used for return values; undefined in JS usually means uninitialized var. Better return null

rszwajko replied:
This is consistent with what we return currently - note that optional chaining operator `?.` and  `find()` return `undefined`.  IMHO having both in one method is overkill.

metalice replied:
?. make sense to return undefined, https://262.ecma-international.org/6.0/#sec-terms-and-definitions-undefined-type , u can also commit the return value and simply use return , u will get same value

rszwajko commented on src/utils/components/InstanceTypeModal/StandaloneInstanceTypeModal.tsx:
```diff
@@ -0,0 +1,36 @@
+
+  return (
+    <InstanceTypeModal
+      {...otherProps}
```
the opinion is not commonly shared:)

metalice commented:
/lgtm
